### PR TITLE
ssl: support enabling TLSv1.3 via lua_ssl_protocols

### PR DESCRIPTION
**Pull request by ghedo**
This simply mirrors commit https://github.com/nginx/nginx/commit/9a37eb3a62130473596e0e4c2e388d80bdb14956#diff-0584d16332cf0d6dd9adb990a3c76a0c in NGINX.

I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

### CODE REVIEW

**Comment by agentzh** [comment]
@ghedo Is it possible to add a test case to the test suite so that we can have this change properly covered?

**Comment by agentzh** [comment]
@ghedo Thanks for your contribution as always!

**Comment by ghedo** [comment]
The current TLSv1.3 situation is such that it would require special NGINX+OpenResty builds (e.g. with BoringSSL, or with OpenSSL 1.1 which is not supported) so I don't think it's currently possible to add tests for this.

**Comment by agentzh** [comment]
@ghedo I see. We need to polish up and merge one of the OpenSSL 1.1 PRs first.

**Comment by agentzh** [comment]
@ghedo OpenSSL 1.1.0 support has landed in the latest master branch. Just FYI :) Thanks!

**Comment by ghedo** [comment]
I added a commit to test this feature, but apparently OpenSSL hasn't released TLSv1.3 in a stable version yet, so the build fails.

**Comment by ghedo** [comment]
And even when OpenSSL will release it, the test will only work in OpenSSL 1.1.0, so the 1.0.2 builds will fail anyway.

**Comment by agentzh** [comment]
@ghedo We can try checking the OpenSSL version number in the `nginx -V` output and skip the whole `.t` test file conditionally (assuming your TLS 1.3 tests are in a separate `.t` file)?

We are already doing something similar in the existing test suite:

https://github.com/openresty/lua-nginx-module/blob/master/t/139-ssl-cert-by.t#L8

Will you update your test file accordingly? Many thanks!

**Comment by ghedo** [comment]
@agentzh ok, done. Since the required OpenSSL version hasn't been release yet, I added a note to the travis config, to add the `enable-tls1_3` option to the OpenSSL build once it gets added, so it won't be forgotten.

**Comment by ghedo** [comment]
Also, for the record, I tested this locally with a git checkout of OpenSSL master and the test works.

File: t/155-tlsv1_3.t [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        listen unix:$TEST_NGINX_HTML_DIR/nginx.sock ssl;
+        server_name   test.com;
+        ssl_certificate ../html/test.crt;
+        ssl_certificate_key ../html/test.key;
+        ssl_protocols TLSv1.2 TLSv1.3;
+
+        server_tokens off;
+        location /foo {
+            default_type 'text/plain';
+            content_by_lua 'ngx.status = 201 ngx.say("foo") ngx.exit(201)';
```
Comment (agentzh) [inline t/155-tlsv1_3.t]: Newly added test cases should use the `*_by_lua_block` directives for inlined Lua code instead.

File: t/155-tlsv1_3.t [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    }
+--- config
+    server_tokens off;
+    lua_ssl_trusted_certificate ../html/test.crt;
+    lua_ssl_protocols TLSv1.2 TLSv1.3;
+    location /t {
+        #set $port 5000;
+        set $port $TEST_NGINX_MEMCACHED_PORT;
+
+        content_by_lua '
```
Comment (agentzh) [inline t/155-tlsv1_3.t]: Ditto.

**Comment by agentzh** [comment]
Just merged with various minor edits and some docs. Thanks!

**Pull request closed without merging**